tf_helpers/lax.py

```diff
@@ -0,0 +1,465 @@
+
+  if (lhs_rank == rhs_rank == 3 and lhs_batch == (0,) and rhs_batch == (0,)
+      and lhs_contraction == (2,) and rhs_contraction == (1,)):
+    return np.asarray(tf.linalg.matmul(lhs, rhs))
+
+  for i in range(len(lhs_contraction)):
+    rhs_rep[rhs_contraction[i]] = lhs_rep[lhs_contraction[i]]
+  for i in range(len(lhs_batch)):
+    rhs_rep[rhs_batch[i]] = lhs_rep[lhs_batch[i]]
+
+  output_rep = _compose_output_rep(lhs_rep, rhs_rep, lhs_contraction,
+                                  rhs_contraction, lhs_batch, rhs_batch)
+  equation = ''.join(lhs_rep) + ',' + ''.join(rhs_rep) + "->" + output_rep
+  return np.asarray(tf.einsum(equation, lhs, rhs))
+
+
+def reduce_window(inputs, init_value, reducer, window_dimensions, strides,
+                  padding, base_dilation=None, window_dilation=None):
+  if reducer not in [np.max, np.add]:
+    raise TypeError("Only max pooling and average/sum pooling are supported.")
+
+  # Note that there is no need to send in the parameter data format since the
+  # input is already of default data format - "N...C". The adjustments of the
+  # input shape is already finished in apply_fun of Pooling in stax.
+  pooling = "AVG" if reducer == np.add else "MAX"
+  output = nn.pool(inputs, window_dimensions, pooling, strides, padding)
+  return np.asarray(output)
+
+
+# TOTO (Zhibo Zhang): Expand the test cases of general convolution and revise
+#                     the according bugs.
+# TODO (Zhibo Zhang): Support feature_group_count, batch_group_count and precision, and
+#       allow lhs_dilation and rhs_dilation to happen at the same time.
+def conv_general_dilated(lhs, rhs, window_strides, padding, output_shape, lhs_dilation=None,
+                         rhs_dilation=None, dimension_numbers=None,
+                         feature_group_count=1, batch_group_count=1, precision=None):
+  """ A general conv API that integrates normal conv, deconvolution,
+  dilated convolution, etc."""
+  # raise TypeError("lhs shape: {}, rhs shape: {}".format(lhs.shape, rhs.shape))
+  dim = None
+  lhs_spec, rhs_spec, out_spec = dimension_numbers
+  if lhs_spec != out_spec:
+    raise TypeError("Current implementation requires the `data_format` of the "
+                    "inputs and outputs to be the same.")
+  if len(lhs_spec) >= 6:
+    raise TypeError("Current implmentation does not support 4 or higher"
+                    "dimensional convolution, but got: ", len(lhs_spec) - 2)
+  dim = len(lhs_spec) - 2
+  if lhs_dilation and rhs_dilation:
+    if lhs_dilation == (1,) * dim and rhs_dilation == (1,) * dim:
+      lhs_dilation, rhs_dilation = None, None
+    else:
+      raise TypeError("Current implementation does not support that deconvolution"
+                    "and dilation to be performed at the same time, but got"
+                    " lhs_dilation: {}, rhs_dilation: {}".format(lhs_dilation,
+                    rhs_dilation))
+  print("the dim is: {}".format(dim))
```

Comment:
Here and elsewhere, is this for debugging? If so, I'd remove all such print statements.